Add reallocation capability to bfc_allocator.

trentlo opened the pull request:
This commit mitigates external fragmentation in bfc_allocator by reallocation.
That is, although the sum of regions and unallocated bytes is larger than the
requested bytes but the bfc_allocator still fails to allocate a large enough
contiguous region to fulfill the request due to fragmentation. To avoid this
case, a relocation feature is implemented to deallocate free regions so that
a larger region can be formed.

Note that essentially we are able to play this trick because the underlying GPU
driver will be able to manipulate the vtable to return a region with a contiguous
virtual address space.

trentlo committed:
Add reallocation capability to bfc_allocator.

This commit mitigates external fragmentation in bfc_allocator by reallocation.
That is, although the sum of regions and unallocated bytes is larger than the
requested bytes but the bfc_allocator still fails to allocate a large enough
contiguous region to fulfill the request due to fragmentation. To avoid this
case, a relocation feature is implemented to deallocate free regions so that
a larger region can be formed.

cheshire commented on tensorflow/core/common_runtime/gpu/gpu_bfc_allocator_test.cc:
```diff
@@ -568,6 +568,47 @@ class GPUBFCAllocatorPrivateMethodsTest : public ::testing::Test {
   }
+
+  void TestRegionDeallocation() {
+    setenv("TF_FORCE_GPU_ALLOW_GROWTH", "unparseable", 1);
```
setenv in tests is really frowned upon if tests execute in parallel, is there a different way?

trentlo replied:
Thanks for bringing up this. I double check. I can remove this line.

But please note that (un)setenv has already been used in this class. E.g., see this [line](https://github.com/tensorflow/tensorflow/blob/30f2dff032a27be0d4adbb06979de6f54df1105e/tensorflow/core/common_runtime/gpu/gpu_bfc_allocator_test.cc#L419). This env variable is unset in the `SetUp` of this test so that line 573 is not necessary. The test behavior we need is simply to test with `allow_growth=true` specified by the `GPUOption`. That is why I wanted to unset the env var `"TF_FORCE_GPU_ALLOW_GROWTH"`.

cheshire commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -260,6 +261,76 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
+  // Searching for free regions.
+  absl::flat_hash_set<void*> free_region_ptrs;
+  size_t total_free_bytes = 0;
+  for (const auto& region : region_manager_.regions()) {
```
I would prefer an explicit type here.

tayo replied:
Accessing region_manager_ in this function requires holding mutex lock_

cheshire commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -260,6 +261,76 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
+    return false;
+  }
+
+  VLOG(INFO) << "Re-allocate memory regions to avoid OOM due to memory"
```
Should this be upgraded to LOG(WARNING)? Running near the threshold of available memory can incur a significant performance slowdown (in my understanding), it might be better to make it more explicit.

tayo replied:
+1

cheshire replied:
I think you want LOG(WARNING), not VLOG(WARNING)

trentlo replied:
Ah. I also just noticed that myself. But good catch. Thanks.

cheshire commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -260,6 +261,76 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
+
+  // Deallocate free regions.
+  auto it = region_manager_.regions().begin();
+  while (it != region_manager_.regions().end()) {
```
Nitpick: IMO it would be more readable to have this loop in a static helper function.

tayo commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -260,6 +261,76 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
+    ChunkHandle h = region_manager_.get_handle(region.ptr());
+    bool any_use = false;
+    while (h != kInvalidChunkHandle) {
+      const Chunk* c = ChunkFromHandle(h);
```
Calling ChunkFromHandle() requires holding mutex lock_

tayo replied:
Also true throughout this function.  Actually this may be best solved by taking the lock at the beginning of this function.

tayo replied:
Agreed.

trentlo reviewed (commented):
Thanks for the review. Some response. Will update the codes accordingly.

trentlo committed:
Some code refactory and polishing.

trentlo commented:
Revised based on the received review comments. Please help to take a look again when you have a moment. Thanks.

trentlo committed:
Minor tweak for the warning message.

tayo commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -295,16 +294,25 @@ bool BFCAllocator::DeallocateFreeRegions(size_t rounded_bytes) {
-             << " that the re-allocation may incur performance overhead despite"
-             << " better memory utilization. You may try smaller batch sizes"
-             << " to see if it can give you better performance.";
+  VLOG(WARNING) << "Re-allocate memory regions (i.e., allocations) to avoid OOM"
```
LOG(WARNING) ?

trentlo replied:
I also just noticed that myself but good catch!

trentlo committed:
Use LOG(WARNING) instead of VLOG(WARNING).

trentlo reviewed (commented):
Thanks for taking a look again. Good catch on the `VLOG`.

cheshire commented:
@tayo do you want to give the final LGTM?

jlebar commented:
I think it would be wise to add a guard flag so that if this causes problems it can be disabled?  (Unless that's already there and I'm missing it.)

jlebar commented:
If I'm reading the code correctly, this only helps when allow-growth is on, is that right?  Because we can only free units malloc'ed by individual cudaMalloc calls (these are called "regions" in the code?).  But when allow-growth is off, we malloc one big region, so the only case when we're able to free it is when it's empty, but then there's no need to free it.

trentlo commented:
> If I'm reading the code correctly, this only helps when allow-growth is on, is that right? Because we can only free units malloc'ed by individual cudaMalloc calls (these are called "regions" in the code?). But when allow-growth is off, we malloc one big region, so the only case when we're able to free it is when it's empty, but then there's no need to free it.

That is a correct read. This commit only helps the mode with `allow_growth=true`, as in the `allow_growth=false` model the entire memory is just one region (yes, this is a malloc'ed allocation) in the current implementation.

Some data points: I hit this memory segmentation issue with a resent50 model from _Joy of Cooking_ and it has `allow_growh=true` by default in its implementation. Without the PR, we can only bring up the fp32 model up to batch_size=96 w/ XLA, while vanilla TF can run with batch_size=128. With this PR, XLA can run up to batch_size=160. Somehow surprisingly, even if I set `allow_growth=false`, XLA cannot bring up the model with batch_size=128.

One more thing I'd like to point out for discussions is that it is possible for this PR to help with the `allow_growth=false` case, as long as we pre-allocate the entire memory with a set of regions with sizes, e.g., 100MB+200MB+400MB+...+4GB+8GB (instead of one giant region). I did not make the corresponding change because it is not (yet) clear to me if it will hurt some cases.

trentlo commented:
> I think it would be wise to add a guard flag so that if this causes problems it can be disabled? (Unless that's already there and I'm missing it.)

I don't mind and can help to add a guard flag (by introducing an env variable to turn it off?). However, logically, the new feature is only triggered when OOM happens in the original implementation. So, turning off the feature can only hit an OOM when there is an issue executing the new reallocation feature.

That's been said, it may still be possible that having a guard flag can help with debugging? Let me know what you think.

jlebar commented:
> I hit this memory fragmentation issue with a resent50 model from Joy of Cooking and it has allow_growh=true by default in its implementation.

Perhaps this explains some or all of the difference between what you were seeing (lots of OOMs) and what I was seeing (few OOMs) when we last talked about this in our meeting.  allow_growth=true is not the default setting in TensorFlow overall.

> However, logically, the new feature is only triggered when OOM happens in the original implementation. So, turning off the feature can only hit an OOM when there is an issue executing the new reallocation feature.

I think some users may prefer to know that they're OOMing instead of hitting potentially pathological slowdowns in their code.  A guard flag would also aid in debugging if we suspect this change could be causing an unexpected problem.  In general guard flags are cheap and can have large benefits.

trentlo commented:
>> ... Somehow surprisingly, even if I set allow_growth=false, XLA cannot bring up the model with batch_size=128.

> Perhaps this explains some or all of the difference between what you were seeing (lots of OOMs) and what I was seeing (few OOMs) when we last talked about this in our meeting. allow_growth=true is not the default setting in TensorFlow overall.

Note that I meant (as quoted above) that in my observed case, setting `allow_growth=false` does not help. Only `allow_growth=true` with this PR can bring up the ResNet50 with batch_size not smaller than vanilla TF can do.

> I think some users may prefer to know that they're OOMing instead of hitting potentially pathological slowdowns in their code. A guard flag would also aid in debugging if we suspect this change could be causing an unexpected problem. In general guard flags are cheap and can have large benefits.

Will add a guard flag. (Hard to say what users want.)

jlebar commented:
> Note that I meant (as quoted above) that in my observed case, setting allow_growth=false does not help.

Maybe I misunderstand, but I thought you wrote above that with allow_growth=False XLA gets batch 128.  With allow_growth=True and without this PR, XLA gets batch 96.

I don't mean to parse words, but that would suggest to me that allow_growth=False *does* help, just not as much as you might want?

I agree it is odd that you are able to match TF's batch 160 with allow_growth=True + this PR but not with allow_growth=False.  I don't have an explanation for that, but I think it might be worth trying to understand.

tayo commented:
> @tayo do you want to give the final LGTM?

Will do after guard flag commit, requested by jlebar

tayo commented:
When allow_growth=false, the allocator's constructor begins by claiming the entirety of memory.  When allow_growth=true, the allocator starts by claiming 1MB.  

Perhaps starting with one large chunk that is split as needed will more quickly lead to the large-block external fragmentation issue that Trent has observed.  Almost like initializing the state of free memory with a different 'seed'.

trentlo committed:
Add a guard flag for the new garbage collection feature.

trentlo commented:
Just pushed a new commit adding the guard flag. Please help to take a look. @tayo 

I double checked and here are some notes and clarifications:
1. I was not precise. **Both** `allow_growh=false` _and_ `allow_growh=true` with this PR can do batch_size=160 for my Joy-of-Cooking Resnet50.

2. So, @jlebar you are right that this explains _partial_ difference between our observations about OOM.

3. We do have another XLA OOM bug (Joy-of-Cooking GNMT) with `allow_growth=false` due to _internal_ memory fragmentation.

tayo reviewed (dismissed):
This looks good to me.

tayo commented:
> 3\. We do have another XLA OOM bug (Joy-of-Cooking GNMT) with `allow_growth=false` due to _internal_ memory fragmentation.

FYI, there is a max internal fragmentation parameter you could play with, in case you were not aware.
https://github.com/tensorflow/tensorflow/blob/83f4a0c638988a040824081d0224964e1684214a/tensorflow/core/common_runtime/bfc_allocator.cc#L445

jlebar commented:
> I was not precise. Both allow_growh=false and allow_growh=true with this
PR can do batch_size=160 for my Joy-of-Cooking Resnet50.

That is surprising, since we said that this change should have no effect
with allow_growth=false, right?  Do we understand this?

On Fri, Jul 19, 2019, 12:46 PM Tayo Oguntebi <notifications@github.com>
wrote:

> 3. We do have another XLA OOM bug (Joy-of-Cooking GNMT) with
> allow_growth=false due to *internal* memory fragmentation.
>
> FYI, there is a max internal fragmentation parameter you could play with,
> in case you were not aware.
>
> https://github.com/tensorflow/tensorflow/blob/83f4a0c638988a040824081d0224964e1684214a/tensorflow/core/common_runtime/bfc_allocator.cc#L445
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/30827?email_source=notifications&email_token=AABEZB62CJAEHVEDWMHIXPDQAIKZTA5CNFSM4IEVUDSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2MSKAI#issuecomment-513352961>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABEZB5YNFRXH22KNWMXS2DQAIKZTANCNFSM4IEVUDSA>
> .
>

tayo commented:
> we said that this change should have no effect with allow_growth=false

I did not agree with this.  This change will have an effect with allow_growth=false.  The value of allow_growth would simply determine the starting state of the free chunk list.  This change takes effect when the OOM is about to occur, by doing the collection pass.

jlebar commented:
Tayo, Trent agreed with it earlier:
https://github.com/tensorflow/tensorflow/pull/30827#issuecomment-513100384
It's surprising to me that he's now saying that this does have an effect
with allow_growth=false, because I don't see how that can matter.  There is
only one allocation to play with (?), so you can't free and coalesce allocs
like this PR tries to.


On Fri, Jul 19, 2019, 1:11 PM Tayo Oguntebi <notifications@github.com>
wrote:

> we said that this change should have no effect with allow_growth=false
>
> I did not agree with this. This change will have an effect with
> allow_growth=false. The value of allow_growth would simply determine the
> starting state of the free chunk list. This change takes effect when the
> OOM is about to occur, by doing the collection pass.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/30827?email_source=notifications&email_token=AABEZBYV44WMA3P2W7FABODQAINVNA5CNFSM4IEVUDSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2MUCXA#issuecomment-513360220>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABEZB34PB2KJ6KX7B36STTQAINVNANCNFSM4IEVUDSA>
> .
>

trentlo commented:
@jlebar Some natural language ambiguous. Please read the following rephrased sentence again.

Both `allow_growh=false without the PR` and `allow_growh=true with this PR` can do batch_size=160 for my Joy-of-Cooking Resnet50.

trentlo commented:
> > 3. We do have another XLA OOM bug (Joy-of-Cooking GNMT) with `allow_growth=false` due to _internal_ memory fragmentation.
> 
> FYI, there is a max internal fragmentation parameter you could play with, in case you were not aware.
> 
> https://github.com/tensorflow/tensorflow/blob/83f4a0c638988a040824081d0224964e1684214a/tensorflow/core/common_runtime/bfc_allocator.cc#L445

Will play with the parameter later. Thanks for the info.

tayo commented:
@jlebar -- Good point

@trentlo -- Can you confirm that with this PR, your runs converge / work as expected?

tayo commented on tensorflow/core/common_runtime/bfc_allocator.cc:
```diff
@@ -260,6 +262,90 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
+  absl::flat_hash_set<void*> free_region_ptrs;
+  size_t total_free_bytes = 0;
+  for (const AllocationRegion& region : region_manager_.regions()) {
+    ChunkHandle h = region_manager_.get_handle(region.ptr());
```
I believe this returns the handle at the beginning of the region. Regions are further divided into bins.

The while loop then iterates over all the chunks in its bin.  If I read this correctly, it does not iterate over all the chunks in the region.  Is this an issue?

trentlo replied:
A `region` contains a linked list of chunks. `Bin` is another data structure recording a set of chunk handles (see `FreeChunkSet`) that have similar sizes to assist search for best fit. So, for each `region`, we can visit all the chunks (owned by the region) through the linked list.

The for loop at line 274 iterates through all the `regions`. For each `region`, we then iterate through all the `chunks` in the `region` following the linked list (i.e., the while loop at line 277).

Checking all the chunks in a region is enough to verify whether the region is free or in use. When we de-allocate the region, we have to remove all its chunks from Bins (line 340.) This makes sure no dangling chunk handles in Bins.

trentlo committed:
Fixing a compilation error.

Some compilers disallow passing const_iterator to std::vector::erase()
(while some allow).

trentlo commented:
> @trentlo -- Can you confirm that with this PR, your runs converge / work as expected?

Yes, they work as expected. It is some natural language ambiguity as I replied [here](https://github.com/tensorflow/tensorflow/pull/30827#issuecomment-513371892).

BTW, I just pushed a commit fixing a compilation error observed in the (previous) CI runs. Some compilers do not allow passing const_iterator to std::vector::erase(). It will be good if you can take a look.

jlebar commented:
Sorry, I am still having difficulty assimilating everything that was said here into a consistent picture of the world.

> Note that I meant (as quoted above) that in my observed case, setting allow_growth=false does not help. Only allow_growth=true with this PR brings up the ResNet50 with batch_size not smaller than vanilla TF can do.

This seems unambiguous to me.  But now I'm hearing something which seems to contradict this:

> Both allow_growh=false without the PR and allow_growh=true with this PR can do batch_size=160 for my Joy-of-Cooking Resnet50.

I would assume that the first was simply an error, but I am hearing that this is all due to natural language ambiguity, so I am not sure, maybe I'm misunderstanding something here?

Perhaps in the name of clarity we could make the table specifying the max batch size for all combinations of the following booleans?

 * TF classic vs XLA
 * allow_growth = true vs false
 * without this PR vs with this PR

trentlo commented:
> Sorry, I am still having difficulty assimilating everything that was said here into a consistent picture of the world.

No problem. Let's see if we can make things clearer.


> 
> > Note that I meant (as quoted above) that in my observed case, setting allow_growth=false does not help. Only allow_growth=true with this PR brings up the ResNet50 with batch_size not smaller than vanilla TF can do.

I realized this above statement was wrong after I double checked the experiments. Setting `allow_growth=false` XLA can do batch_size=160 (PR is irrelevant in this case). Sorry for the confusion.


> 
> > Both allow_growh=false without the PR and allow_growh=true with this PR can do batch_size=160 for my Joy-of-Cooking Resnet50.
> 
> Perhaps in the name of clarity we could make the table specifying the max batch size for all combinations of the following booleans?
> 
>     * TF classic vs XLA
> 
>     * allow_growth = true vs false
> 
>     * without this PR vs with this PR

In the Joy-of-Cooking Resnet50 case,
TF classic: batch_size=160 (PR irrelevant in this case; allow_growth irrelevant too.)
XLA `allow_growth=false`: batch_size=160 (PR irrelevant in this case)
XLA `allow_growth=true` _without_ this PR: batch_size=96
XLA `allow_growth=true` _with_ this PR: batch_size=160

Let me know if this clarifies the discussions.

jlebar commented:
> Let me know if this clarifies the discussions

Yes, thanks a lot!

skye commented:
Thanks for the reviews everyone, I think you got this under control :) Removing myself as a reviewer.
